src/ccl_fftlog.c

```diff
@@ -140,83 +141,126 @@ static void fht(int npk, int N,
+                                dim, mu, q, kcrc, u, status, \
+                                forward_plan, reverse_plan, \
+                                L, ulocal)
+    {
+      int local_status = 0;
+
+      double *prefac_pk=NULL;
+      if(local_status == 0) {
+        prefac_pk = malloc(N*sizeof(double));
+        if(prefac_pk==NULL)
+          local_status=CCL_ERROR_MEMORY;
+      }
 
-  if(*status == 0) {
-    for(int i = 0; i < N; i++)
-      prefac_pk[i] = pow(k[i], dim/2-q);
+      double *prefac_xi=NULL;
+      if(local_status == 0) {
+        prefac_xi = malloc(N*sizeof(double));
+        if(prefac_xi==NULL)
+          local_status=CCL_ERROR_MEMORY;
+      }
 
-    /* Compute k's corresponding to input r's */
-    double k0r0 = kcrc * exp(-L);
-    r[0] = k0r0/k[0];
-    for(int n = 1; n < N; n++)
-      r[n] = r[0] * exp(n*L/N);
+      fftw_complex* a=NULL;
+      fftw_complex* b=NULL;
+      if(local_status == 0) {
+        a = fftw_alloc_complex(N);
+        if(a==NULL)
+        local_status=CCL_ERROR_MEMORY;
+      }
 
-    double one_over_2pi_dhalf = pow(2*M_PI,-dim/2);
-    for(int i = 0; i < N; i++)
-      prefac_xi[i] = one_over_2pi_dhalf * pow(r[i], -dim/2-q);
+      if(local_status == 0) {
+        b = fftw_alloc_complex(N);
+        if(b==NULL)
+          local_status=CCL_ERROR_MEMORY;
+      }
 
-    /* Compute the convolution b = a*u using FFTs */
-    fftw_plan forward_plan = fftw_plan_dft_1d(N, (fftw_complex*) a, (fftw_complex*) b,
-					      -1, FFTW_ESTIMATE);
-    fftw_plan reverse_plan = fftw_plan_dft_1d(N, (fftw_complex*) b, (fftw_complex*) b,
-					      +1, FFTW_ESTIMATE);
-
-    for(int j = 0; j < npk; j++) {
-      for(int i = 0; i < N; i++)
-	a[i] = prefac_pk[i] * pk[j][i];
-
-      fftw_execute(forward_plan);
-      for(int m = 0; m < N; m++)
-	b[m] *= u[m] / (double)(N);       // divide by N since FFTW doesn't normalize the inverse FFT
-      fftw_execute(reverse_plan);
-
-      /* Reverse b array */
-      double complex tmp;
-      for(int n = 0; n < N/2; n++) {
-	tmp = b[n];
-	b[n] = b[N-n-1];
-	b[N-n-1] = tmp;
+      if(local_status == 0) {
+        for(int i = 0; i < N; i++)
+          prefac_pk[i] = pow(k[i], dim/2-q);
+
+        /* Compute k's corresponding to input r's */
+        double k0r0 = kcrc * exp(-L);
+        r[0] = k0r0/k[0];
+        for(int n = 1; n < N; n++)
+          r[n] = r[0] * exp(n*L/N);
+
+        double one_over_2pi_dhalf = pow(2*M_PI,-dim/2);
+        for(int i = 0; i < N; i++)
+          prefac_xi[i] = one_over_2pi_dhalf * pow(r[i], -dim/2-q);
+
+        #pragma omp for
+        for(int j = 0; j < npk; j++) {
+          for(int i = 0; i < N; i++)
+            a[i] = prefac_pk[i] * pk[j][i];
+
+          fftw_execute_dft(forward_plan,a,b);
+          for(int m = 0; m < N; m++)
+            b[m] *= u[m] / (double)(N);       // divide by N since FFTW doesn't normalize the inverse FFT
+          fftw_execute_dft(reverse_plan,b,b);
```

Comment:
Is it safe to reuse plans from multiple threads?

Reply:
Yes (in fact having plans be initialized within parallel regions is a no-no)
http://www.fftw.org/fftw3_doc/Thread-safety.html